### PR TITLE
restrict partnership_admin access to calendars by neighbourhood AND partnership tag

### LINKED ISSUE #2170: Partnership admins can only view calendars that belong to partners they can view
## Acceptance criteria

- [ ] Partnership admins can only see calendars that belong to the partners they can view
- [ ] Partnership admins cannot see calendars that belong to partners that are only in their neighbourhood scope
- [ ] Partnership admins cannot see calendars that belong to partners that are only in their tag scope

### PR DESCRIPTION
**Pull request by aaaaargZombies**
fixes #2170 

- adjust scope to account for tags
- adjust edit policy to prevent partnership_admins accessing calendars directly via URL

### CODE REVIEW

File: app/policies/calendar_policy.rb [hunk 1]
```diff
@@ -14,15 +14,12 @@ def new?
...
+    update?
   end
 
   def update?
     return true if user.root?
-    return true if user.partner_admin? && user.partner_ids.include?(record.partner_id)
-
-    # return true if user.neighbourhood_admin? && user.neighbourhoods.include?(record.address.neighbourhood)
-    index?
+    return true if scope.map(&:id).include? record.id
```
Comment (r-ferrier) [inline app/policies/calendar_policy.rb:22]: oh this is so much simpler! Why don't we always do this!

File: app/policies/calendar_policy.rb [hunk 2]
```diff
@@ -42,16 +39,32 @@ def resolve
       return scope.all if user.root?
       return scope.none if !user.partner_admin? && !user.neighbourhood_admin?
```
Comment (r-ferrier) [inline app/policies/calendar_policy.rb:40]: don't you need to add in a `!user.partnership_admin?` check here too?

File: app/policies/calendar_policy.rb [hunk 3]
```diff
@@ -42,16 +39,32 @@ def resolve
...
-                  'service_areas.neighbourhood_id in (:neighbourhood_ids) OR '\
-                  'calendars.partner_id IN (:partner_ids))',
-                  neighbourhood_ids: user.owned_neighbourhood_ids,
-                  partner_ids: user.partner_ids
-                )
-                .distinct
+      if user.partnership_admin?
+        user_partnership_tag_ids = user.tags.map(&:id)
+        partnership_calendars =
+          Calendar.left_joins(partner: %i[address service_areas partnerships])
```
Comment (r-ferrier) [inline app/policies/calendar_policy.rb:45]: nice!

**Review by r-ferrier — CHANGES_REQUESTED** [review]
One change (maybe) needed?

File: app/policies/calendar_policy.rb (hunk 2 above)
Comment (aaaaargZombies) [reply app/policies/calendar_policy.rb]: yes I think this is a good catch, all the conditionals are getting a bit hairy round here

**Review by r-ferrier — APPROVED** [review]

**Pull request merged**